README.md
### Sigma Designs
### Author: Brian Han
### Date Created: July 10, 2017
### Last Updated: August 11, 2017

# README

- This repo contains all the test scripts at my time at Sigma Designs.
- These scripts automate testing LTE for Sigma chips, using Litepoint
as the primary testing system, with other components integrated in later.
- "WIP" is where the various iterations of my scripts are located, including
the latest version and the scripts described in the "Documentation for Relevant
Scripts" section below.
- "shortened" contains a working GUI and testing script; is the
precursor to the following...
- "test_MiniUT_PCBA_ATE" is the version of the GUI and testing script sent out.
- If you wish to trace the progress of the script over time, view
the CHANGELOG.md located in the "WIP" folder.
- "Sample-Code" is the starter code provided by Sirius Ding at
Litepoint to help get started working with the Litepoint machine.
-"pyFunctionalTest" contains example code using pySerial to help
me get started using the serial library.

-------------------------------------------------------------------------------
-------------------------------------------------------------------------------
-------------------------------------------------------------------------------

# Setup

- These files are to be run using Python 2.7.13.
Doing otherwise would use the wrong socket library conventions.
(ASCII v Binary)
- Third-Party Python Library Modules Needed for these programs:
        PySerial
        PyVisa
        ftd2xx
- In addition, to PyVisa, you will need to install NI-VISA from 
National Instruments, as well as their drivers for GBIB - USB.
The driver version used for these programs were NI-488.2

### More Detailed Setup 

For more on setup and how to run these programs,
 checkout the 'README.md' located in the
"test MiniUT PCBA ATS" folder. It will show you how to run the latest version
of the testing script, as well as all the installations needed first.


-------------------------------------------------------------------------------
-------------------------------------------------------------------------------
-------------------------------------------------------------------------------

# Documentation for Relevant Scripts

## test_miniUT_v2_shortened.py

This is the latest version of the testing script. The test sweeps nRB and
gain. miniUT.setup file is the configuration file for this script. The logic
for this script is as follows:
- Find the GPIB port used for the power supply (Agilent 6631B) and set voltage
to 2.5, and the current to low power.
- Find the GPIB port used for the multimeter (Fluke 8845A)
- Find the current at low power mode, communicating with the power supply,
relay, and the multimeter to make the current measurement.
- Setup a connection to Litepoint and the DUT, and perform the sweep
of nRB and gain. The results of each individual test is stored into a .csv
file, generated in the current directory.
- Run the specification check to make sure the results of the sweep match
the desired range.

## MiniUT_GUI_v2_shortened.py

This is the GUI for the test_miniUT_v2_shortened.py script. The GUI
expects user to enter three key inputs:
- The HOST IP address to establish a connection to Litepoint
- The CABLE LOSS in dB from the cable to properly calculate the average
power level
- The SN (Serial Number) of the board in question.

The SN can be typed in, or scanned in by barcode. Once satisfiable inputs
are entered, the program can be run with the click of a button. In addition
to just key values, there are some features in the menubar the user can use:
- A standard "Exit" option
- A "Set Default Values" option so the user can set what values the default
button will set the key values to
- A "Settings" option that allows the user to customize their output options.
- An "Update Firmware" option that allows the user to update the firmware of
the board.
> NOTE: The "Update Firmware" feature is commented out. It works, but you will
need to uncomment the logic for updating the firmware

## test_miniUT_sweep_five_values.py

> NOTE: While the latest script, test_miniUT_v2_shortened.py has logic to
automatically find the correct ports for the measuring equipment and the COM
for the DUT, this script does not contain that logic, and these connections
must be set using global variables found at the top of the script.
> POWER SUPPLY: This script uses the E3648A power supply, not the 66311B. The
SCPI commands to program E3648A are not the same for 66311B (check out
test_66311B.py and test_3648A.py to see the differences).

This script is the one used for overnight testing (as it takes upwards of
7 - 8 hours), as it sweeps 5 sets of values: The PA High Power / Low Power mde,
The PA Bias, The PA VCC, nRB, and gain. The script sweeps the values in that 
order, gain being the innermost value swept first.

> NOTE: This script does not take advantage of the configuration file
miniUT.setup. This script uses "input_spec.txt" instead to get the specification
range to compare the results to.

## simple_RB_PA_test.py

- This is a basic script you can use to quickly set up the DUT and the Litepoint
to measure the TX (it does not actually run the test; you do it youself through
Litepoint's provided GUI)
- Use it to do a quick unit test.
- The script DOES NOT set up the CSW (crystal) or automate the power supply
- You can customize the DUT setup with the global variables at the top of the
script.

## setup_generation.py

- This is used to generate the miniUT.setup file used by test_miniUT_v2_shortened
to set up the test.
- There are directions in the file on how to read it and make adjustments.
- Included in case of accidental destruction of the current miniUT.setup file 

WIP/CHANGELOG.md
## Sigma
## Author: Brian Han
## Date: 8/10/17

# Changelog

All version changes from file to file is documented here.
Files are listed from newest to oldest order, descending by section

***************************************************************************
***************************************************************************
***************************************************************************

# GUIs

## MiniUT_GUI_v2_shortened.py
### Added
- Menu bar, with 4 features: exit, set default values, settings,
update firmware
- exit is self-explanatory
- set default values allow user to set what the default button sets the
default values as
- settings enable special debugging features and print statements
- update firmware updates the firmware on the board; this feature is
currently commented out; uncomment to use

### Change
- Small adjustments made to the GUI appearance


## MiniUT_GUI_shortened
### Added
- Runs script test_miniUT_shortened.py
- Button resets input values to default
- Button resets input values to previous test values
- CABLE LOSS and HOST input added to account for loss in dB and getting
the connection to the Litepoint machine respectively

## MiniUT_GUI_Black_with_subprocess_old_copy
### Added
- Update of MiniUT_GUI_Black

### Changed
- Use subprocessing to run script without GUI hanging
- Runs script test_miniUT_shortened.py
- Improved regex logic

## test_miniUT_GUI_bad_thread.py
### DOES NOT WORK

## flash_login_miniUT_GUI
### INCOMPLETE; USE "flash_miniUT_GUI" INSTEAD

## flash_miniUT_GUI
### Added
- Flash the DUT using Serial library
- User defines the path to a firmware file with a file dialog
- Update DUT firmware with the press of a button
- Implement threading to prevent GUI from crashing

### Changed
- Different functionality from the previous GUI; not an update

## MiniUT_GUI_Black
### Added 
- Error Message Handling
- Regex logic to make sure user has not entered garbage values

### Changed 
- Runs test_miniUT_RB_PS_66311B_copy_9
- User enters only the Serial Number and the COM for the DUT

## LTE_GUI_Basic
### Added
- First version of the miniUT GUI
- Allow user to enable debugging, set gain range, enter
serial number, set voltage, and set current
- Runs test_miniUT_RB_PS_66311B with the inputed settings
- Note that this version is not multithreaded or multiprocessed,
so program hangs while the script is running
- Not recommended for use

***************************************************************************
***************************************************************************
***************************************************************************

# Test Scripts

## test_miniUT_v2_shortened
### Added
- Latest Version of Test Script
- 

### Changed
- Specification check put back in
- Reads more values from miniUT.setup

## test_miniUT_shortened
### Added
- 'low power mode' test is properly implemented
- Implemented test_sweep_0_1f logic for finding CSW
- Now reads some values from miniUT.setup

### Changed
- Test is reduced back to only sweeping RB then gain

### Removed
- Specification Check is commented out, waiting until values are finalized.

## test_miniUT_sweep_five_values
### Changed
- Implemented a cleaner way to write the .csv file

### Removed
- No longer setting fine gain with 'wr 242 4444'

## modified_test_miniUT_HH_RB_PA_E3648A
### REDACTED SCRIPT; DO NOT USE

## test_miniUT_crystal_rb
### REDACTED SCRIPT; DO NOT USE

### Changed
- Not a direct upgrade over the previous script; serves a different purpose

## test_miniUT_RB_PA_E3648A
### IMPORTANT NOTE
- The 'low power mode' measurement is not implemented correctly
in this version. If you want to use the 5-range sweep, comment out
line #735, and run the script 

### Added
- Sweep 5 sets of values: PA Mode (high or low power), 
PA Bias (0,10,20,30,7f, fe), PA VCC (1 to 3.5 V), nRB, gain
- Measure voltage at 'low power mode' (unimplemented unit testing)
- Logic to communicate with relay and Fluke 8845A multimeter
- Boolean to enable logging of all output to the console
- Logic to retry measurement a few times in case of signal loss

### Changed
- Using power supply E3648A

## test_miniUT_RB_PS_E3648A
### Changed
- Using power supply E3648A
- Go back to creating one .csv file for all nRB and their gain sweep

## test_miniUT_RB_PS_66311B
### Added
- Boolean to enable break-like steps at the end of each measurement
- Error message printing and return values
- Test to compare the output with expected values from input_spec.txt 

### Changed
- Creates separate .csv files for each nRB with their gain values
- Using power supply 66311B
- Integrated test_crystal logic into the main test script

### Removed
- Dependence on test_crystal to find CSW

## test_miniUT_gain_rb_version_with_curr
### Added
- Make call to test_crystal.main() to find CSW

### Changed
- Sweeps gain 0,1,2,3,4,5,10,20...70

### Removed
- Ask user input for gain range to sweep
- No longer in high power mode 

## test_gain_rb_version_with_curr
### Added
- Measuring current and voltage in addition to the other measurements
- Turn off power supply when the test finishes

## test__rb_and_rb_offset_version
### Added
- Sweep nRB, RB offset, and gain
- Asks user for RB offset
- Enable / Disable some print statements with a DEBUG variable
- Turn on Power Supply E3648A to set Voltage and current
- Sweep nRB, RB offset, and gain
- Set fine gain using 'wr 242 4444'

## test_gain_rb_version
### Added
- Sweeps nRB, then gain
- Set gain using 'd 26 X'
- Asks user for gain range to sweep
- Delay between sending Litepoint calculation command and requesting
the result
- Capture ACLR E-UTRA Lower and Upper in the measurements

### Changed
- Measurements are now done for 10 subframes, not 1, taking the average

WIP/json_practice.py
#Practice place for JSON file interaction
#DictReader/DictWriter v json
#Will need some overhead to switch current data and logic from
#dict to json
#With json, when I read the data from the file, they are 
#automatically formatted into a data structure that I can use.
#Currently, I have to do this step myself, which becomes
#more complicated when I require nested data structures
#e.g. rb_hex
#Should think of switching to json NOW
# JSON to Python object chart:
#   JSON            Python
#   object          dict
#   array           list
#   string          str
#   number(int)     int
#   number(real)    float
#   true            True
#   false           False
#   null            None



import json
"""
info = {"menu": {
  "id": "file",
  "value": "File",
  "popup": {
    "menuitem": [
      {"value": "New", "onclick": "CreateNewDoc()"},
      {"value": "Open", "onclick": "OpenDoc()"},
      {"value": "Close", "onclick": "CloseDoc()"}
    ]
  }
}} 
with open ("miniUT.setup", 'wb') as file:
    json.dump (info, file) 
    file.close()

with open ("miniUT.setup", 'rb') as file:
    data = json.load(file)
    print type(data)
    print data["menu"]["popup"]["menuitem"][0]["value"]
"""

"""
        SN
        DUT
        CSW_CENTER (e.g. 9 out of 0 - 1f)
        COM, GPIB (? Have logic to make it a dynamic check in test_Visa)
        PS_NAME (important later, maybe if using multiple power supplies)
        HOST
        PORT (maybe not)
"""

info = {

        "VOLT" : 2.5,
        "HOST" : '10.10.14.202',
        "PORT": '24000',
        "CSW_CENTER": "9",
        "DUT": "miniUT Rev 8",
        "SN":"10",
        "COM": "COM8",
        "PS_ADDRESS": 'GPIB::9::INSTR',
        "CABLE_LOSS_DB": 11,
        "EVM_LIMIT": 6,
        "DEBUG": False,
        "STEP_TEST": False,
        "LOGGING": True,
        "rb_and_scale": [
            ['1', '97A0'],
            ["24",'177F'],
            ["50", '1758']
        ],
        "rb_and_scale_key": {
            
            1: '97A0',
            2: '9772',
            3: '57B6',
            4: '579E',
            5: '578D',
            6: '5781',
            8: '5770',
            9: '576A',
            10:'5765',
            12:'17B4',
            15:'17A1',
            16:'179C',
            18:'1793',
            20:'178B',
            24:'177F',
            25:'177D',
            27:'1778',
            30:'1772',
            32:'176E',
            36:'1768',
            40:'1763',
            45:'175D',
            48:'175A',
            50:'1758',
        },
        "specs": {
            "POWER_LOWER_LIMIT" : -60,
            "POWER_UPPER_LIMIT": 30,
            "IQ_OFFSET_LOWER_LIMIT": -40,
            "IQ_OFFSET_UPPER_LIMIT": -20,
            "FREQ_ERROR_LOWER_LIMIT": -3000,
            "FREQ_ERROR_UPPER_LIMIT": 3000,
            "DATA_EVM_LOWER_LIMIT": 2,
            "DATA_EVM_UPPER_LIMIT": 30,
            "PEAK_DATA_EVM_LOWER_LIMIT": 10,
            "PEAK_DATA_EVM_UPPER_LIMIT": 70,
            "RS_EVM_LOWER_LIMIT": 2,
            "RS_EVM_UPPER_LIMIT": 30,
            "PEAK_RS_EVM_LOWER_LIMIT": 10,
            "PEAK_RS_EVM_UPPER_LIMIT": 500,
            "IQ_IMBALANCE_GAIN_LOWER_LIMIT": -1,
            "IQ_IMBALANCE_GAIN_UPPER_LIMIT": 1,
            "IQ_IMBALANCE_PHASE_LOWER_LIMIT": -1,
            "IQ_IMBALANCE_PHASE_UPPER_LIMIT": 1,
            "ACLR_EUTRA_L_LOWER_LIMIT": 20,
            "ACLR_EUTRA_L_UPPER_LIMIT": 70,
            "ACLR_EUTRA_U_LOWER_LIMIT": 20,
            "ACLR_EUTRA_U_UPPER_LIMIT": 70,
            "CURRENT_LOWER_LIMIT": 0,
            "CURRENT_UPPER_LIMIT": 2
        }
    }

with open ("miniUT.setup", 'wb') as file:
    json.dump (info, file) 
    file.close()


with open ("miniUT.setup", 'rb') as file:
    data = json.load(file)
    print type(data)
    print data["rb_and_scale"][0][0]
    print data["rb_and_scale_key"]["1"]
    print data["SN"]
    """
    if '48' in data["rb_and_scale_key"].keys():
        data["rb_and_scale"].append(['48',data["rb_and_scale_key"]['48']])
    """
    file.close()

with open ("miniUT.setup", 'wb') as file:
    json.dump (data, file)
    file.close()
